scripts/ci/render_bench.py

```diff
@@ -0,0 +1,380 @@
+    benchmarks: Benchmarks = {}
+
+    def insert(name: str, entry: BenchmarkEntry) -> None:
+        if name not in benchmarks:
+            benchmarks[name] = []
+        benchmarks[name].append(entry)
+
+    for date, commits in commits_by_date.items():
+        for commit in commits:
+            short_sha = commit[0:7]
+            path = short_sha_to_path(short_sha)
+            if path not in bucket:
+                continue
+            for entry in entry_parser(bucket[path].download_as_text()):
+                name = entry["name"]
+                entry["date"] = f"{date.year}-{date.month}-{date.day}"
+                entry["commit"] = short_sha
+                insert(name, entry)
+            break
+
+    return benchmarks
+
+
+def get_crates_benchmark_data(gcs: storage.Client, commits_by_date: CommitsByDate) -> Benchmarks:
+    print('Fetching benchmark data for "Rust Crates"…')
+
+    return collect_benchmark_data(
+        commits_by_date,
+        bucket=fetch_bucket(gcs, "rerun-builds", "benches"),
+        short_sha_to_path=lambda short_sha: f"benches/{short_sha}",
+        entry_parser=lambda data: [parse_bencher_line(line) for line in non_empty_lines(data)],
+    )
+
+
+def get_size_benchmark_data(gcs: storage.Client, commits_by_date: CommitsByDate) -> Benchmarks:
+    print('Fetching benchmark data for "Sizes"…')
+
+    return collect_benchmark_data(
+        commits_by_date,
+        bucket=fetch_bucket(gcs, "rerun-builds", "sizes/commit"),
+        short_sha_to_path=lambda short_sha: f"sizes/commit/{short_sha}/data.json",
+        entry_parser=parse_sizes_json,
+    )
+
+
+BYTE_UNITS = ["b", "kb", "kib", "mb", "mib", "gb", "gib", "tb", "tib"]
+VALID_CONVERSIONS = {
+    "ns/iter": ["ns/iter"],
+}
+for unit in BYTE_UNITS:
+    VALID_CONVERSIONS[unit] = BYTE_UNITS
+
+UNITS = {
+    "b": 1,
+    "kb": 1000,
+    "kib": 1024,
+    "mb": 1000,
+    "mib": 1024 * 1024,
+    "gb": 1000,
+    "gib": 1024 * 1024,
+    "tb": 1000,
+    "tib": 1024 * 1024,
+    "ns/iter": 1,
+}
+
+
+def convert(base_unit: str, unit: str, value: float) -> float:
+    """Convert `value` from `base_unit` to `unit`."""
+    base_unit = base_unit.lower()
+    unit = unit.lower()
+    if unit not in VALID_CONVERSIONS[base_unit]:
+        raise Exception(f"invalid conversion from {base_unit} to {unit}")
+    return value / UNITS[unit] * UNITS[base_unit]
+
+
+def min_and_max(data: list[float]) -> (float, float):
+    min_value = float("inf")
+    max_value = float("-inf")
+    for value in data:
+        if value < min_value:
+            min_value = value
+        if value > max_value:
+            max_value = value
+    return (min_value, max_value)
+
+
+def render_html(title: str, benchmarks: Benchmarks) -> str:
+    print(f'Rendering "{title}" benchmark…')
+
+    chartjs = {}
+    for name, benchmark in benchmarks.items():
+        if len(benchmark) == 0:
+            chartjs[name] = None
+        labels = [entry["date"] for entry in benchmark]
```

Comment:
please add the commit hash for every entry as well. Date is nice to get a feel for things and set the x-axis, but won't help much to find issues :)

Reply:
Also needs an onclick handler which links to the commit.

Reply:
uh yeah that'd be really cool <3